sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedPartitions.scala

```diff
@@ -0,0 +1,253 @@
+          getMapStartIndices(right, partitionId, rightMedSize)
+        } else {
+          Array(0)
+        }
+
+        if (leftMapIdStartIndices.length > 1 || rightMapIdStartIndices.length > 1) {
+          skewedPartitions += partitionId
+          for (i <- 0 until leftMapIdStartIndices.length;
+               j <- 0 until rightMapIdStartIndices.length) {
+            val leftEndMapId = if (i == leftMapIdStartIndices.length - 1) {
+              getNumMappers(left)
+            } else {
+              leftMapIdStartIndices(i + 1)
+            }
+            val rightEndMapId = if (j == rightMapIdStartIndices.length - 1) {
+              getNumMappers(right)
+            } else {
+              rightMapIdStartIndices(j + 1)
+            }
+            // TODO: we may can optimize the sort merge join to broad cast join after
+            //       obtaining the raw data size of per partition,
+            val leftSkewedReader =
+            SkewedShufflePartitionReader(
+              left, partitionId, leftMapIdStartIndices(i), leftEndMapId)
+
+            val rightSkewedReader =
+              SkewedShufflePartitionReader(right, partitionId,
+                rightMapIdStartIndices(j), rightEndMapId)
+            subJoins += SortMergeJoinExec(leftKeys, rightKeys, joinType, condition,
+              leftSkewedReader, rightSkewedReader)
+          }
+        }
+      }
+      logDebug(s"number of skewed partitions is ${skewedPartitions.size}")
+      if (skewedPartitions.size > 0) {
+        val optimizedSmj = smj.transformDown {
+          case sort: SortExec if (
+            ShuffleQueryStageExec.isShuffleQueryStageExec(sort.child)) => {
+            val newStage = ShuffleQueryStageExec.getShuffleStage(
+              sort.child.asInstanceOf[QueryStageExec]).copy(
+              excludedPartitions = skewedPartitions.toSet)
+            sort.copy(child = newStage)
+          }
+        }
+        subJoins += optimizedSmj
+        UnionExec(subJoins)
+      } else {
+        smj
+      }
+  }
+
+  override def apply(plan: SparkPlan): SparkPlan = {
+    if (!conf.getConf(SQLConf.ADAPTIVE_EXECUTION_SKEWED_JOIN_ENABLED)) {
+      return plan
+    }
+
+    def collectShuffleStages(plan: SparkPlan): Seq[ShuffleQueryStageExec] = plan match {
+      case _: LocalShuffleReaderExec => Nil
+      case _: CoalescedShuffleReaderExec => Nil
+      case stage: ShuffleQueryStageExec => Seq(stage)
+      case ReusedQueryStageExec(_, stage: ShuffleQueryStageExec, _) => Seq(stage)
+      case _ => plan.children.flatMap(collectShuffleStages)
+    }
+
+    val shuffleStages = collectShuffleStages(plan)
+
+    if (shuffleStages.distinct.length == 2) {
+      // Currently we only support handling skewed join for 2 table join.
+      handleSkewJoin(plan)
+    } else {
+      plan
+
+    }
+  }
+}
+
+/**
+ * A wrapper of shuffle query stage, which submits one reduce task to read the partition produced
```

Comment:
`which submits one task to read shuffle partitions for reducer 'QueryStageExec' produced by mappers in range ...`